Scrollable frame

dorileo opened the pull request:
This PR includes a small fix, a theme change and a new feature for OnKeyDown events callback calls and more importantly a new feature for frame control. We can now make it scrollable.

@VladimirMarkelov can you please help me in depth reviewing the "frame: scrollable frame" patch? this patch touches very sensible areas and I faced many corner cases while implementing it, I suppose you may find a few more. Can you also test with you current test cases for frame?

Thanks...

Commit:
add custom theme support

With this change users can specify custom theme extensions.

Signed-off-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>

Commit:
fix getLinearControlList() fn callback call

If no fn callback is provided don't try to call it and avoid segfault.

Signed-off-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>

Commit:
frame: scrollable frame

This patch introduce internal changes in order to #1 actually clip
content avoid growing parent's size and keeping controls with their
own pre-set sizes and also #2 add a scrollable feature to frame control.

Signed-off-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>

Commit:
add context to OnKeyDown

In order to keep caller's context this patch introduces a interface{}
argument to carry on arbitrary data and pass it back based on the
registered callback.

Signed-off-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>

VladimirMarkelov commented:
A very good one. Thank you! It is long, so it takes some time for me to read and understand. Please be patient ;)

VladimirMarkelov commented on window.go:
```diff
@@ -21,8 +21,14 @@ type Window struct {
+}
+
+type keyDownCb struct {
+	data interface{}
```
Question: how is `data` going to be used? In the patch I do not see an example that explains its usage.

dorileo replied:
data is used by the callback user, i.e:

```go
type UserData struct {
...
}
myData := &UserData{}
wnd.wnd.OnKeyDown(func(evt Event, data interface{}) bool{
     md := data.(UserData)
     // do some data manipulation here this could be some key handling logic i.e alt+<key>
    // manipulation where the first hit detects alt then do some special handling after that
   // and so on
}, myData)

```

VladimirMarkelov commented on theme.go:
```diff
@@ -510,9 +510,11 @@ func ThemeInfo(name string) ThemeDesc {
 // clr - current object color
 // id - color ID in theme
-func RealColor(clr term.Attribute, id string) term.Attribute {
+func RealColor(clr term.Attribute, style string, id string) term.Attribute {
```
Nitpick: I think function comment should be updated to explain new argument

dorileo replied:
sure

VladimirMarkelov commented on theme.go:
```diff
@@ -510,9 +510,11 @@ func ThemeInfo(name string) ThemeDesc {
 // id - color ID in theme
-func RealColor(clr term.Attribute, id string) term.Attribute {
+func RealColor(clr term.Attribute, style string, id string) term.Attribute {
+	ccolor := fmt.Sprintf("%s%s", style, id)
```
Just a question: Can you provide an example of `style`? I see they are just concatenated. Is a point between them make a resulting string better: `ccolor := fmt.Sprintf("%s.%s", style, id)`?

dorileo replied:
sure, I can provide some examples. I like the suggestion of changing it to ```<style>.<id>```

VladimirMarkelov commented on frame.go:
```diff
@@ -67,16 +79,52 @@ func (f *Frame) Draw() {
+		DrawScrollBar(x+w, y, 1, h, f.lastScrollProp)
+	}
+
+	fg, bg := RealColor(f.fg, f.Style(), ColorViewText), RealColor(f.bg, f.style, ColorViewBack)
```
Why one uses `f.Style()` while another one `f.style`?

dorileo replied:
No reason at all, just a careless and inconsistent programmer :)

VladimirMarkelov commented on ctrlutil.go:
```diff
@@ -266,8 +266,59 @@ func SendEventToChild(parent Control, ev Event) bool {
+	xx, yy := c.Pos()
+	ww, hh := c.Size()
+
+	return xx >= x && xx >= x &&
```
`xx >= x && xx >= x` looks suspicious ;)

dorileo replied:
another careless programmer speaking here :)

VladimirMarkelov reviewed (commented):
I looked through the changes but have not tried them yet.

VladimirMarkelov commented:
I tried the patch - applications work fine. So, I'll wait for your changes ad we can merge it ;)

VladimirMarkelov commented:
And it would be great to have a small example of how to use scrolling frames.

Commit:
demo: add a scrollable frame demo

With this demo we show how to use the scrollable frame with a frame
containing 10 buttons.

Signed-off-by: Leandro Dorileo <leandro.maciel.dorileo@intel.com>

dorileo commented:
@VladimirMarkelov updated the set with the requested changes + a patch introducing a small demo with a scrollable frame.
